tests/Feature/Api/V1/GetUserGameLeaderboardsTest.php

```diff
@@ -0,0 +1,406 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Feature\Api\V1;
+
+use App\Models\Game;
+use App\Models\Leaderboard;
+use App\Models\LeaderboardEntry;
+use App\Models\System;
+use App\Models\User;
+use App\Platform\Enums\ValueFormat;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Illuminate\Support\Carbon;
+use Tests\TestCase;
+
+class GetUserGameLeaderboardsTest extends TestCase
+{
+    use RefreshDatabase;
+    use BootstrapsApiV1;
+
+    public function testItValidates(): void
+    {
+        $this->get($this->apiUrl('GetUserGameLeaderboards'))
+            ->assertJsonValidationErrors([
+                'i',
+                'u',
+            ]);
+    }
+
+    public function testGetUserGameLeaderboardsUserNotFound(): void
+    {
+        $this->get($this->apiUrl('GetUserGameLeaderboards', ['i' => 99999, 'u' => 'nonExistant']))
+            ->assertNotFound()
+            ->assertJson(['User not found']);
+    }
+
+    public function testGetUserGameLeaderboardsGameNotFound(): void
+    {
+        /** @var User $user */
+        $user = User::factory()->create();
+
+        $this->get($this->apiUrl('GetUserGameLeaderboards', ['i' => 99999, 'u' => $user->User]))
+            ->assertNotFound()
+            ->assertJson(['Game not found']);
+    }
+
+    public function testGetUserGameLeaderboardsGameHasNoLeaderboards(): void
+    {
+        /** @var User $user */
+        $user = User::factory()->create();
+
+        /** @var System $system */
+        $system = System::factory()->create();
+
+        /** @var Game $game */
+        $game = Game::factory()->create(['ConsoleID' => $system->ID]);
+
+        $this->get($this->apiUrl('GetUserGameLeaderboards', ['i' => $game->ID, 'u' => $user->User]))
+            ->assertUnprocessable()
+            ->assertJson(['Game has no leaderboards']);
+    }
+
+    public function testGetUserGameLeaderboardsUserHasNoLeaderboardsOnGame(): void
+    {
+        /** @var User $user */
+        $user = User::factory()->create();
+
+        /** @var System $system */
+        $system = System::factory()->create();
+
+        /** @var Game $game */
+        $game = Game::factory()->create(['ConsoleID' => $system->ID]);
+
+        Leaderboard::factory()->create(['GameID' => $game->ID]);
+
+        $this->get($this->apiUrl('GetUserGameLeaderboards', ['i' => $game->ID, 'u' => $user->User]))
+            ->assertUnprocessable()
+            ->assertJson(['User has no leaderboards on this game']);
+    }
+
+    public function testGetUserGameLeaderboards(): void
```

Comment:
IMO, there's too much happening in this test case. Ideally, each individual test case should test a single thing and be easy for future maintainers to reason about.

I propose splitting this into separate test cases that each verify a specific behavior:
1. A basic test for a user with a single leaderboard entry.
2. A test for a user with multiple leaderboard entries.
3. Separate pagination tests (first page, last page, partial results, etc.)
4. A test specifically for rank calculation.

These focused test cases would make the test suite more maintainable, improve DX, and make it immediately clear which functionality is broken when a test fails. It would also make it easier to add new test cases in the future should any logic or behavior change.

Historically, in the past we crammed a lot of logic into single test cases because we were fearful of long test execution times. More recently, we've been trending towards caring much less about execution times and instead favoring more bite-sized test cases.

Reply:
I agree, was turning to be complex even to write them. I have to say I also prefer segregation, I'll adjust these. Thanks for the suggestions, clearly helpful :)